x/ibc/callbacks/keeper/keeper.go

```diff
@@ -0,0 +1,157 @@
+	contractAddress string,
+	version string,
+) error {
+	data, err := transfertypes.UnmarshalPacketData(packet.GetData(), version, "")
+	if err != nil {
+		return err
+	}
+
+	// can only call callback if the receiver is the module address
+	receiver := sdk.MustAccAddressFromBech32(data.Receiver)
+	if !receiver.Equals(k.authKeeper.GetModuleAddress(types.ModuleName)) {
+		return nil
+	}
+
+	cbData, isCbPacket, err := callbacktypes.GetCallbackData(data, version, packet.GetDestPort(), cachedCtx.GasMeter().GasRemaining(), cachedCtx.GasMeter().GasRemaining(), callbacktypes.DestinationCallbackKey)
+	if err != nil {
+		return err
+	}
+	if !isCbPacket {
+		return nil
+	}
+
+	ethReceiver := common.BytesToAddress(receiver)
+	contractAddr := common.HexToAddress(contractAddress)
+
+	// TODO: Do something with the response
+	_, err = k.evmKeeper.CallEVMWithData(cachedCtx, ethReceiver, &contractAddr, cbData.Calldata, true)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k ContractKeeper) IBCOnAcknowledgementPacketCallback(
+	cachedCtx sdk.Context,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+	contractAddress,
+	packetSenderAddress string,
+	version string,
+) error {
+	data, _, err := k.packetDataUnmarshaler.UnmarshalPacketData(cachedCtx, packet.GetSourcePort(), packet.GetSourceChannel(), packet.GetData())
+	if err != nil {
+		return err
+	}
+
+	cbData, isCbPacket, err := callbacktypes.GetCallbackData(data, version, packet.GetSourcePort(), cachedCtx.GasMeter().GasRemaining(), cachedCtx.GasMeter().GasRemaining(), callbacktypes.SourceCallbackKey)
+	if err != nil {
+		return err
+	}
+	if !isCbPacket {
+		return nil
+	}
+
+	sender := common.BytesToAddress(sdk.MustAccAddressFromBech32(packetSenderAddress))
+	contractAddr := common.HexToAddress(contractAddress)
+
+	// TODO: Do something with the response
+	_, err = k.evmKeeper.CallEVMWithData(cachedCtx, sender, &contractAddr, cbData.Calldata, true)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k ContractKeeper) IBCOnTimeoutPacketCallback(
+	cachedCtx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+	contractAddress,
+	packetSenderAddress string,
+	version string,
+) error {
+	data, _, err := k.packetDataUnmarshaler.UnmarshalPacketData(cachedCtx, packet.GetSourcePort(), packet.GetSourceChannel(), packet.GetData())
+	if err != nil {
+		return err
+	}
+
+	cbData, isCbPacket, err := callbacktypes.GetCallbackData(data, version, packet.GetSourcePort(), cachedCtx.GasMeter().GasRemaining(), cachedCtx.GasMeter().GasRemaining(), callbacktypes.SourceCallbackKey)
+	if err != nil {
+		return err
+	}
+	if !isCbPacket {
+		return nil
+	}
+
+	sender := common.BytesToAddress(sdk.MustAccAddressFromBech32(packetSenderAddress))
+	contractAddr := common.HexToAddress(contractAddress)
+
+	// TODO: Do something with the response
+	_, err = k.evmKeeper.CallEVMWithData(cachedCtx, sender, &contractAddr, cbData.Calldata, true)
```

Comment:
If we support different success and failure calldata for acknowledgement, we can use the failure calldata here as well.